### PR TITLE
Bodrov Daniil. Lab 3. Var 2

### PR DESCRIPTION
**Pull request by overinvest**
In this pull request, I have implemented a pass that counts the number of machine instructions performed during the execution of a function, excluding the increment of the instruction counter. 

Please review the changes and let me know if you have any questions or suggestions. Thank you!

### CODE REVIEW

File: llvm/lib/Target/X86/X86TargetMachine.cpp [hunk 1]
```diff
@@ -659,4 +659,4 @@ bool X86PassConfig::addRegAssignAndRewriteOptimized() {
     addPass(createX86TileConfigPass());
   }
   return TargetPassConfig::addRegAssignAndRewriteOptimized();
-}
+}
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86TargetMachine.cpp]:
```suggestion
}

```

File: llvm/lib/Target/X86/lab3/CMakeLists.txt [hunk 2]
```diff
@@ -10,4 +10,4 @@ if (NOT WIN32 AND NOT CYGWIN)
   ENDFOREACH()
 
   set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
-endif()
+endif()
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/CMakeLists.txt]:
```suggestion
endif()

```

File: llvm/lib/Target/X86/lab3/bodrov_daniil/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h)
+add_llvm_pass_plugin(${PluginName} 
+  ${ALL_SOURCE_FILES}     
+  DEPENDS
+  intrinsics_gen
+  X86
+  BUILDTREE_ONLY)
+
+target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bodrov_daniil/CMakeLists.txt]:
```suggestion
target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})

set(LLVM_TEST_DEPENDS ${PluginName} ${LLVM_TEST_DEPENDS} PARENT_SCOPE)

```

File: llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  // Iterate over all basic blocks in the function
+  for (auto &MBB : MF) {
+    unsigned InstructionCount =
+        0; // Counter to store the number of instructions
+    for (auto &MI : MBB) {
+      // Increment the instruction count, excluding the instruction to increment
+      // the counter itself
+      if (MI.getOpcode() !=
+          X86::MOV32ri) { // Exclude the instruction to increment the counter
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp]: There may be another valid instruction `MOV32ri`. And I don't see where you uses such instruction for counter processing.

File: llvm/test/lab3/bodrov_daniil/test.mir [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+
+# void empty_func() {}
+
+# int main() {
+#     int a = 10, b = 100, c = 5;
+#     int res = func(a, b, c);
+#     std::cout << "Number of machine instructions executed for func("
+#               << a << ", " << b << ", " << c << ") - " << ic << std::endl;
+#     return 0;
+# }
```
Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir]:
```suggestion
```
Please remove this and iostream inclusion to simplify MIR code.

File: llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                           GlobalValue::ExternalLinkage, nullptr, "ic");
+  }
+
+  // Get the target instruction info
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+
+  DebugLoc DL3 = MF.front().begin()->getDebugLoc();
+
+  // Get the register for storing the counter
+  unsigned icReg = X86::R10;
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp]: In real program this register may be user by some other instructions and its value may be modified. Please use the register which is allocated for `ic` var directly.

File: llvm/test/lab3/bodrov_daniil/test.mir [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+    renamable $eax = nsw ADD32rr killed renamable $eax, killed renamable $edi, implicit-def dead $eflags
+    renamable $ecx = DEC32r killed renamable $ecx, implicit-def $eflags
+    ; CHECK: $r10 = ADD64ri32 $r10, 5, implicit-def $eflags
+    JCC_1 %bb.5, 5, implicit $eflags
+  
+  bb.6.for.cond.cleanup:
+    liveins: $eax
+  
+    ; CHECK: $r10 = ADD64ri32 $r10, 1, implicit-def $eflags
+    RET64 $eax
```
Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir:491]: Oops, you returned from a function before storing counter back into `ic`.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/bodrov_daniil/test.mir [hunk 8]
```diff
@@ -0,0 +1,878 @@
...
+constants:       []
+machineFunctionInfo: {}
+body:             |
+  bb.0.entry:
+    successors: %bb.2(0x50000000), %bb.1(0x30000000)
+    liveins: $edi, $edx, $esi
+  
+    MOV64mi32 $rip, 1, $noreg, @ic, $noreg, 0 :: (store (s64) into @ic, !tbaa !5)
+    TEST32rr renamable $edi, renamable $edi, implicit-def $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 3, implicit-def $eflags
```
Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir]: Could you please add check of a neighbor instruction to clarify where your new instruction is placed in a code?

Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir]: Could you please add check of a neighbor instruction to clarify where your new instruction is placed in a code?

File: llvm/test/lab3/bodrov_daniil/test.mir [hunk 9]
```diff
@@ -0,0 +1,878 @@
...
+    liveins: $eax
+  
+    ; CHECK: MOV64mr %0, &ic
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 1, implicit-def $eflags
+    RET64 $eax
+  
+  bb.1:
+    ; CHECK: MOV64mr %0, &ic
+    renamable $eax = XOR32rr undef $eax, undef $eax, implicit-def dead $eflags
+    ; CHECK: %0:gr32 = ADD64ri32 %0, 2, implicit-def $eflags
```
Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir]: So, you first move counter value from register into memory and after it increment the counter again?

File: llvm/test/lab3/bodrov_daniil/test.mir [hunk 10]
```diff
@@ -0,0 +1,878 @@
...
+    %sub = sub i32 %add, %z
+    %0 = add i32 %y, -1
+    %1 = mul i32 %sub, %0
+    %2 = add i32 %add, %1
+    %3 = sub i32 %2, %z
+    %min.iters.check = icmp ult i32 %x, 8
+    br i1 %min.iters.check, label %for.cond1.preheader.preheader, label %vector.ph
+  
+  vector.ph:                                        ; preds = %for.cond1.preheader.lr.ph
+    %4 = icmp sgt i32 %y, 0
```
Comment (m-ly4) [inline llvm/test/lab3/bodrov_daniil/test.mir]: Please remove llvm ir comment part.

File: llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  DebugLoc DL3 = DebugLoc(); // Empty DebugLoc for simplicity
+
+  // Iterate over all basic blocks in the function
+  for (auto &MBB : MF) {
+    unsigned InstructionCount =
+        0; // Counter to store the number of instructions
+    for (auto &MI : MBB) {
+      // Check if the instruction is not the one used for incrementing the
+      // counter
+      if (MI.getOpcode() != X86::ADD32ri && MI.getOpcode() != X86::ADD64ri32) {
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bodrov_daniil/X86CountInstBD.cpp]: Again, there may be other add64ri32 instructions that should be counted.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by overinvest** [comment]
```
for (auto &MI : MBB) {
      // Check if the instruction modifies or reads the EFLAGS register
      if (MI.modifiesRegister(X86::EFLAGS, TRI) ||
          MI.readsRegister(X86::EFLAGS, TRI)) {
        if (MI.isCompare()) {
          InsertPt = &MI;
        }
      }
    }
```

This snippet iterates through the instructions in a basic block. When it encounters an instruction that modifies or reads the flags register (EFLAGS), it checks if the opcode of the instruction corresponds to a CMP instruction `X86::CMP...` . If so, it updates the insertion point to point to this CMP instruction.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**